### DISCUSSION TITLE
Visualizing streamlines without streamplot.

### DISCUSSION THREAD

**Discussion by quenot** [magpylib/magpylib#677] [General]
The [Matplotlib Streamplot](https://magpylib.readthedocs.io/en/latest/_pages/gallery/gallery_vis_mpl_streamplot.html#gallery-vis-mpl-streamplot) page of the documentation illustrates how to display streamlines using matplotlib's streamplot function. However, I wanted to display unbroken streamlines and [I struggled to do it with streamplot](https://stackoverflow.com/questions/77344931/how-can-i-avoid-spiraling-streamlines-in-matplotlib-pyplot-streamplot) because streamplot does not handle well closed streamlines. After [asking for support from matplotlib developpers](https://github.com/matplotlib/matplotlib/issues/27261), it turned out that using streamplot is not the most practical way to do that. Instead, it is much easier to [build a stream function from the field](https://math.stackexchange.com/a/1784352/1221716) and to display its contours. I have provided an example derived from one of those provided on the Matplotlib Streamplot page and I think it might be useful for magpylib users: https://stackoverflow.com/a/77419388/13707125

**Comment by OrtnerMichael** [comment]
hi @quenot,

This is really nice!!! Vector field analysis ftw!!!

I followed your references and try to reproduce the argument here in a compressed form for other readers.

![image](https://github.com/magpylib/magpylib/assets/47384012/4702be32-fbe7-41a3-96bf-925717bf7aeb)

This shows that the contours of the stream function PSI correspond to the stream lines of a 2D field u if div(u)=0. If u is a magnetostatic field that is the result of homogeneously magnetized bodies (Magpylib case) then div(u)=0 holds everywhere except at the magnet boundaries, where the field will not be evaluated in any case.

Then you compute the function PSI at all grid points by using `np.cumsum` which beautifully synergizes with the problem as is computes "all integrals" at the same time.

Here is a slightly modified version of the core of your code:

```
# Create an observer grid in the xz-symmetry plane
ts = np.linspace(-5,5,100)
grid = np.array([[(x,0,z) for x in ts] for z in ts]) # slow Python loop

# Compute the B-field of a cube magnet on the grid
cube = magpy.magnet.Cuboid(magnetization=(0, 0, -1000), dimension=(2, 2, 2))
B = cube.getB(grid)
x, y, u, v = grid[:,:,0], grid[:,:,2], B[:, :, 0], B[:, :, 2],

# Build the stream function and plot its isolines
dx, dy = x[0,1]-x[0,0], y[1,0]-y[0,0]
psi = np.cumsum(u,axis=0)*dy-np.cumsum(v[0:1,:],axis=1)*dx
psi -= np.min(psi)
levels = np.sort([psi[50,ii] for ii in range(45,55)]) # select points to draw contour lines from
plt.contour(x, y, psi, levels=levels, colors='blue', linewidths=1.0)
```

which produces something like this:

![image](https://github.com/magpylib/magpylib/assets/47384012/9b3965a1-38dd-443b-9b8a-fe0063b6b8c0)

**Some questions**

Since you are integrating from always the same start position I would expect the integration path to somehow cross the respective magnet boundaries where divB!=0, and possibly to produce faulty results thereon. I did not manage to create faulty results. The algorithm seems to be extraordinary stable. Do you know why?

Have you considered what is the influence of discretization? I noticed that for similar contour levels, when increasing the mesh finesse, the results are quite different. The figure shows for a discretization of 40 (blue) 80(red) 160(green). Probably should not be surprised that the contour levels are moving....but then there is this puzzling left-right asymmetry. Any thoughts on this?

![image](https://github.com/magpylib/magpylib/assets/47384012/5d047f6e-26b6-4446-bcaa-d3c42a996f5d)

Do you know what happens when you are not in a symmetry plane?

@peter-leitner and @fslanovc : Please feel free to join the discussion! Any insights would be appreciated.

Once all issues are understood, I would love to add this to the Magpylib examples!!! As you brought this up @quenot, would you like to add this by yourself, which would also automatically make you a Magpylib contributor. I'd create a documentation page template in a separate branch for you to work in. In case you don't know how, I'd be happy to guide you through the process. In case you don't want to, I would take care of it.

Thanks for bringing this to our attention!

BR Michael

**Reply by quenot** [reply]
Hi Michael,

Warning, edited multiples times.

Great summary. Indeed, the method has nothing to do with incompressible flows. Only the fact that div **u** = 0 matters.

Yes, I understand that there are some limitations that have to be mentioned: the div2D B = 0 and the grid resolution.

Regarding the first one, div B = 0 is guaranteed in 3D by the non existence of magnetic monopoles but we are in a 2D slice here and we have div2D B = 0 if and only if dB/dz everywhere in the plane (assuming an x-y plane), which is guaranteed if the plane in a symmetry one and is possibly hard to be obtained otherwise. If there is no symmetry, it is likely that B has at least in some places non-zero z components and that the streamlines would no longer be planar. So what we would see as streamlines from 2D restriction of the field would not actually be streamlines and probably not even projections of these. However, it may be that not far away from a symmetry plane, the z components of the field and its first derivative are negligible and that we would still see something meaningful. In this case, for taking into account the fact that we only have div2D D ~ 0, the integration would have to be as symmetrical as possible. I would suggest taking the origin in the center of the figure and integrate twice, once x first and then y second and vice versa, and to average the results. Anyway, integrating the 2D restriction with streamplot would also not have a good physical interpretation either. Only streamlines for fields with div2D B = 0 are likely to be strictly meaningful.

Regarding the grid resolution, indeed, the result will depend on the resolution, presumably both for the stream function and the streamline integration method (and there is a step scale parameter additionally hidden in the latter). Maybe a good approach would be to try different resolutions as you just did and increase the resolution until some stability is obtained. Fair comparisons between the approaches could be done in this way but I did non try yet. As I mentioned, obtaining complete unbroken streamlines with streamplot is cumbersome.

Regarding the asymmetry, I also observed it and it is why I switch from from a 200x200 grid to a 201x201 one where it happens to be reduced (I am not sure why). I suspect that it is due to some asymmetry in the cumsum operation leading to non-symmetric heads and tails. In the beginning, many small values are accumulated with a good relative precision while at the end these small values are added to a bigger base or anyway to a quantity to which larger rounding errors have already been included. Probably determining the levels with something symmetrical relatively to the center value (which has to be interpolated for grids with even widths and/or heights) would restore the symmetry. Another possibility is to place the origin in the center of the figure and have cumsum run in both directions (which is a bit tricky for grids with even widths and/or heights). I guess that centering the levels of the lines around the figure center psi value is simpler and as efficient.

Regarding the question of div2D B actually not being 0 at the magnet boundary and its effect on the computation, I am not sure about how to interpret it. It seems that this does not hurt but it is not completely clear why. Here are my thoughts. Magnetic monopoles are not supposed to exist and my understanding is that the computation does "as if" by considering a surface distribution of them as a limit case of a volume distribution of dipoles. However, this is not a "Dirac-like" distribution whose volume integration would lead to non-zero finite quantities. In other words, there is a region where div B is non-zero and even infinite. but it is of null measure and that may be why it does not hurt. At these boundaries, the psi function remains C0 continuous but it is not C1 continuous (as a result of the integration of a function with a discontinuous step) and that must be why we see sharp changes of direction in the psi isolines there. This is certainly fuzzy and partly incorrect, and maybe a professional mathematician could help here. I am afraid that I am not expert enough to propose much better. After thinking about it a bit more, maybe more correctly: there may be a true "dirac-like" distribution that leads to non-zero and even infinite values for div B on a surface of null 3D measure. A first integration gives B with just a step-type discontinuity there and then a second integration gives psi as C1 everywhere except on the "charged" boundaries where it is C0 but not C1. I would say that this is not expected to be surprising or problematic.

I might directly take care of the new example page on magpylib though this is a quite modest contribution. I would probably need some guidance for working with github for this.

I am also wondering: as I understood that B is computed from closed form analytical expressions, is there a chance that psi could be directly computed in a similar way?

Best regards,

Georges.

**Reply by OrtnerMichael** [reply]
hi @quenot 

**Concerning the asymmetry:** I actually do not find it. The asymmetry I observe is for even and odd discretization not there when being very careful about choosing the levels. For example: for splitting into 200 steps, the levels psi[99] and psi[100] are symmetric, while the same levels are not symmetric when splitting into 201 steps. This is ofc trivial but might be overlooked. Do you have an example where you have asymmetry that is not the result of what I just described?

**Concerning the magnet boundaries**: Probably you are right about that.

**div2D B=0**: In the symmetry plane Bz=0, also Bz(z) must be symmetric (NOT antisymmetric), and physics demands that dBz/dz is continuous. This means that Bz(z) must be an extremum which "proofs" that d Bz/dz =0 ... i think this makes sense.

**discretization** I agree. The computation of psi is done numerically, and that integral can only be as precise as the integration which depends on the level splitting. In the example should probably be referred to.

**whats also nice**: This will ofc not just work in Matpotlib but also in Plotly and any other library.

**analytical expression**: Yes Magpylib relies on purely analytical expressions. It might be possible to solve the integral, however, most Magpylib expressions are already so complicated that it might be a considerable effort. And you have to repeat the effort for all different shapes ... And then the probably very long outcome has to be properly implemented (vectorization, avoiding numerical instabilities .... ) ... not sure if its worth it.

I suggest we transfer everything concerning the docu example to #678 .

**Reply by quenot** [reply]
Hi Michael,
Regarding the div2D B question, I think I am still confused about it but when I look at the streamlines plot:
![image](https://i.stack.imgur.com/tUcju.png)
and also at the numerical values in the neighborhood, it seems that we have a sharp variation of the direction of By when x varies and a "Dirac" on dBY/dx at the boundary but possibly a continuity for dBx/dx and dBy/dy and possibly a null sum of them. I am not sure if this is correct but, as dBy/dx does not affect the divergence, we may still have div2D B = 0 everywhere, including the boundary and its neighborhood.
This hurts my previous understanding that div2D B is non-zero for regions containing charges, though.

**Reply by quenot** [reply]
@OrtnerMichael I had a look at the paper "Calculation of the magnetic stray field of a uniaxial magnetic domain" by R. Engel-Herbert and T. Hesjedal that you refereed to and I think that we are safe with div B = 0 absolutely everywhere, including the boundaries. There is a "dipole density" with a step but no magnetic charge. We do have "a Dirac" on dBy/dx at the boundary but this has no effect on the divergence as dBx/dx and dBy/dy are continuous and summing to zero.

**Comment by OrtnerMichael** [comment]
@quenot has prepared a nice code for 2D field lines. You can find it at [his repository](https://github.com/quenot/closestreamlines).

The code is based on the method proposed by @Alexboiboi in #678, which uses pyvista streamtubes in 2D. The major issue is the integration error so that stream tubes are not properly closing, but @quenot found a good way to compensate this error.
